src/grpc/projects.rs

```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
-        //         Object::add_key_value(&dataset_id, client, kv)
-        //             .await
-        //             .map_err(|e| {
-        //                 log::error!("{}", e);
-        //                 tonic::Status::aborted("Database transaction error.")
-        //             })?;
-        //     }
-        // } else if !inner_request.remove_key_values.is_empty() {
-        //     let rm_kv: KeyValues = inner_request.remove_key_values.try_into().map_err(|e| {
-        //         log::error!("{}", e);
-        //         tonic::Status::internal("KeyValue conversion error.")
-        //     })?;
-        //     let object = Object::get(dataset_id, client)
-        //         .await
-        //         .map_err(|e| {
-        //             log::error!("{}", e);
-        //             tonic::Status::aborted("Database transaction error.")
-        //         })?
-        //         .ok_or(tonic::Status::invalid_argument("Dataset does not exist."))?;
-        //     for kv in rm_kv.0 {
-        //         object.remove_key_value(client, kv).await.map_err(|e| {
-        //             log::error!("{}", e);
-        //             tonic::Status::aborted("Database transaction error.")
-        //         })?;
-        //     }
-        // } else {
-        //     return Err(tonic::Status::invalid_argument(
-        //         "Both add_key_values and remove_key_values empty.",
-        //     ));
-        // }
-
-        // let dataset_with_relations = Object::get_object_with_relations(&dataset_id, client)
-        //     .await
-        //     .map_err(|e| {
-        //         log::error!("{}", e);
-        //         tonic::Status::aborted("Database transaction error.")
-        //     })?;
-        // let project = Some(dataset_with_relations.try_into().map_err(|e| {
-        //     log::error!("{}", e);
-        //     tonic::Status::internal("Dataset conversion error.")
-        // })?);
-
-        // Ok(tonic::Response::new(UpdateProjectKeyValuesResponse {
-        //     project,
-        // }))
+        log_received!(&request);
+
+        let token = tonic_auth!(
+            get_token_from_md(request.metadata()),
+            "Token authentication error."
+        );
+
+        let request = KeyValueUpdate::Project(request.into_inner());
+        let collection_id = tonic_invalid!(request.get_id(), "Invalid project id");
+        let ctx = Context::res_ctx(collection_id, DbPermissionLevel::WRITE, true);
+
+        tonic_auth!(
+            self.authorizer.check_permissions(&token, vec![ctx]).await,
+            "Unauthorized"
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_keyvals(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
```

Comment:
Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::WRITE, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        let project = tonic_internal!(
            self.database_handler.update_keyvals(request).await,
            "Internal database error."
        );
        self.cache
            .update_object(&project.object.id, project.clone());
        let project: generic_resource::Resource =
            tonic_internal!(project.try_into(), "Project conversion error");
```